### DISCUSSION TITLE
Single continuous tube mesh with varying radii

### DISCUSSION THREAD

**Discussion by jaylee321** [pyvista/pyvista#4300] [Q&A]
Hi, I'm trying to make a mesh of a blood vessel using PolyDataFilters.tube. My input data consists of two txt files. One file has x,y,z coordinates of nodes and their corresponding radii, and the second file contains connection data between the nodes. My current version of a tubular mesh consists of multiple smaller, disconnected tubes, which is not what I want. Also, I included the scalars array to vary the tube radii but I would not like the final mesh to be colored. How can I make a single mesh that looks like the grey mesh in the picture below?
![idealmesh](https://user-images.githubusercontent.com/131219743/232945139-a1502321-aeb8-452b-9ee0-d9aaf78a8510.jpg)
The code I used and the resulting window is shown here:
```python
import numpy as np
import pyvista as pv

points = np.loadtxt("Data/example_centerline_ICA_nds.txt")
connectivity = np.loadtxt("Data/example_centerline_ICA_edg1.txt")

def lines_from_points(points):
    poly = pv.PolyData()
    poly.points = points[:,:3]
    cells = np.full((len(points), 3), 2, dtype=np.int_)
    cells[:, 1] = connectivity[:,0]
    cells[:, 2] = connectivity[:,1]
    poly.lines = cells
    return poly

line = lines_from_points(points)

min_radius = np.amin(points[:,3])
max_radius = np.amax(points[:,3])
line['scalars'] = points[:,3]

tube = line.tube(radius=min_radius, scalars='scalars', radius_factor=max_radius/min_radius)
tube.plot(opacity=0.5)
```
![pyvistademo](https://user-images.githubusercontent.com/131219743/232947519-b84e8e62-0392-40cb-abe6-1d7b6fe23764.jpg)

**Comment by banesullivan** [comment]
This problem has come up a number of times: individual line cells in a polyline will result in these separated/segmented tubes. The trick is to create a single polyline cell to have one continuous tube

It looks like you are using `lines_from_points` from the polyline example: https://docs.pyvista.org/version/stable/examples/00-load/create-spline.html

Try `polyline_from_points` from that same example: 

```py
def polyline_from_points(points):
    poly = pv.PolyData()
    poly.points = points
    the_cell = np.arange(0, len(points), dtype=np.int_)
    the_cell = np.insert(the_cell, 0, len(points))
    poly.lines = the_cell
    return poly
```

Hope this helps!

**Reply by banesullivan** [reply]
Ah I see now that you modified that helper method for your data... will have to do the same here

**Reply by jaylee321** [reply]
Hi @banesullivan , thank you for your input. I apologize if this question had been asked earlier. I tried incorporating your solution in my code. However, while modifying the ```line_from_points``` function was straightforward, I don't understand how I can include my connection data in the ```polyline_from_points``` function. Each row in the connections data (_edg1 file) represents the points (i.e. their row index in the _nds file) that are connected to each other. Any ideas for specifying this order of connections?

```python
import numpy as np
import pyvista as pv

points = np.loadtxt("Data/example_centerline_ICA_nds.txt")
connectivity = np.loadtxt("Data/example_centerline_ICA_edg1.txt").astype(int)

def polyline_from_points(points):
    poly = pv.PolyData()
    poly.points = points[:,:3]
    the_cell = np.arange(0, len(points), dtype=np.int_)
    the_cell = np.insert(the_cell, 0, len(points))
    poly.lines = the_cell
    return poly

line = polyline_from_points(points)

min_radius = np.amin(points[:,3])
max_radius = np.amax(points[:,3])
line['scalars'] = points[:,3]

tube = line.tube(radius=min_radius, scalars='scalars', radius_factor=max_radius/min_radius)
tube.plot(opacity=0.5)
```
<img width="448" alt="pyvistamesh" src="https://user-images.githubusercontent.com/131219743/233353205-cebf6eb6-9a94-44d7-b0c4-c2191055ba9d.PNG">

**Reply by banesullivan** [reply]
Ah hm. I think I lead you astray now

Let's use the `strip()` filter on the line mesh you have before running the `tube()` filter: https://docs.pyvista.org/version/stable/api/core/_autosummary/pyvista.PolyDataFilters.strip.html#strip

```py
tube = line.strip().tube(radius=min_radius, scalars='scalars', radius_factor=max_radius/min_radius)
```

(ignore the polyline stuff as this `strip()` filter will join adject lines)

```py
import numpy as np
import pyvista as pv

points = np.loadtxt("Data/example_centerline_ICA_nds.txt")
connectivity = np.loadtxt("Data/example_centerline_ICA_edg1.txt")

def lines_from_points(points):
    poly = pv.PolyData()
    poly.points = points[:,:3]
    cells = np.full((len(points), 3), 2, dtype=np.int_)
    cells[:, 1] = connectivity[:,0]
    cells[:, 2] = connectivity[:,1]
    poly.lines = cells
    return poly

line = lines_from_points(points)

min_radius = np.amin(points[:,3])
max_radius = np.amax(points[:,3])
line['scalars'] = points[:,3]

tube = line.strip().tube(radius=min_radius, scalars='scalars', radius_factor=max_radius/min_radius)
tube.plot(opacity=0.5)
```
![screenshot (1)](https://user-images.githubusercontent.com/22067021/234031357-14a9c5dc-3e7f-46bc-9a04-5549d155fdb3.png)
